### PR TITLE
chainHead: Propagate results on the `chainHead_follow`

### LINKED ISSUE #1115: Subxt integration tests failed against latest Substrate build.
The nightly CI run which downloads the latest version of Substrate ran into test failures, which likely means that there are breaking changes that need fixing in Subxt.

Go to https://github.com/paritytech/subxt/actions/workflows/nightly.yml to see details about the failure.

### PR DESCRIPTION
**Pull request by lexnv**
After the change introduced in substrate https://github.com/paritytech/substrate/pull/14692, the `chainHead` API propagates the response of `chainHead_body`, `chainHead_call` and `chainHead_storage` on the `chainHead_follow` subscription.

- the `chainHead` related events are updated
- storage related events are removed
- `chainHead_body`, `chainHead_call` and `chainHead_storage` are now methods that return an `operationId`
- `chainHead_follow` produces events with the result of methods identified by `operationId`

Closes https://github.com/paritytech/subxt/issues/1115.

### CODE REVIEW

**Review by jsdw — APPROVED** [review]
Awesome!

File: subxt/src/rpc/types.rs [hunk 1]
```diff
@@ -476,34 +533,101 @@ pub enum FollowEvent<Hash> {
...
+    ///
+    /// Repeating the same operation in the future will not succeed.
+    OperationError(OperationError),
     /// The subscription is dropped and no further events
     /// will be generated.
     Stop,
 }
 
-/// The result of a chain head method.
+/// The storage item received as paramter.
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:560]:
```suggestion
/// The storage item received as parameter.
```

File: subxt/src/rpc/types.rs [hunk 2]
```diff
@@ -452,16 +449,76 @@ pub struct Finalized<Hash> {
...
-/// 3. BestBlockChanged - indicate that the best block
-///      is now the one from this event. The block was
-///      announced priorly with the `NewBlock` event.
+/// 3. BestBlockChanged - indicate that the best block is now the one from this event. The block was
+///    announced priorly with the `NewBlock` event.
 /// 4. Finalized - State the finalized and pruned blocks.
+///
+/// The following events are related to operations:
+/// - OperationBodyDone: The response of the `chianHead_body`
+/// - OperationCallDone: The response of the `chianHead_call`
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:511]:
```suggestion
/// - OperationCallDone: The response of the `chanHead_call`
```

File: subxt/src/rpc/types.rs [hunk 3]
```diff
@@ -452,16 +449,76 @@ pub struct Finalized<Hash> {
...
 /// 2. NewBlock - a new block was added.
-/// 3. BestBlockChanged - indicate that the best block
-///      is now the one from this event. The block was
-///      announced priorly with the `NewBlock` event.
+/// 3. BestBlockChanged - indicate that the best block is now the one from this event. The block was
+///    announced priorly with the `NewBlock` event.
 /// 4. Finalized - State the finalized and pruned blocks.
+///
+/// The following events are related to operations:
+/// - OperationBodyDone: The response of the `chianHead_body`
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:510]:
```suggestion
/// - OperationBodyDone: The response of the `chainHead_body`
```

File: subxt/src/rpc/types.rs [hunk 4]
```diff
@@ -452,16 +449,76 @@ pub struct Finalized<Hash> {
...
+/// - OperationCallDone: The response of the `chianHead_call`
+/// - OperationStorageItems: Items produced by the `chianHead_storage`
+/// - OperationWaitingForContinue: Generated after OperationStorageItems and requires the user to
+///   call `chainHead_continue`
+/// - OperationStorageDone: The `chianHead_storage` method has produced all the results
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:515]:
```suggestion
/// - OperationStorageDone: The `chainHead_storage` method has produced all the results
```

File: subxt/src/rpc/types.rs [hunk 5]
```diff
@@ -476,34 +533,101 @@ pub enum FollowEvent<Hash> {
...
+}
+
+/// The `started` result of a method.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MethodResponseStarted {
+    /// The operation id of the response.
+    pub operation_id: String,
+    /// The number of items from the back of the `chainHead_storage` that have been discarded.
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:628]: The type doesn't implement `Serialize` so this seems needless to me

File: subxt/src/rpc/types.rs [hunk 6]
```diff
@@ -476,34 +533,101 @@ pub enum FollowEvent<Hash> {
...
+    #[serde(default)]
```
Comment (niklasad1) [inline subxt/src/rpc/types.rs:629]: why is serde(default) needed here?  if the field "discarded_items" is missing and it should be deserialized to   Option::None AFAIU.

**Review by niklasad1 — APPROVED** [review]

**Pull request merged**